### PR TITLE
Minimal test and `.pre-commit` CI workflow

### PR DESCRIPTION
**Pull request by griff-rees**
Addresses  #14.

- Facilitates running `pnpm run test` from the root directory to cover tests in 
  - `template_app`
  - `ua_components`
  - Currently requires `rust` to build `wasm-webpack`
- Adds a minimal `.pre-commit-config.yml`
  - Default linting like trailing whitespace
  - Apply `prettier` as a [`local`](https://pre-commit.com/#repository-local-hooks) using the `format` commands in `template_app` and `ua_components`
- CI action to
  - `test` via `pnpm run test` with `rust build`
  - `linter` via `.pre-commit` config 
  - Runs on
    - merge requests to `main`
    - changes to `main`

What it *doesn't* do (yet):

- Anything regarding `rust` (other than build as necessary to run `template_app` tests)

### CODE REVIEW

**Comment by griff-rees** [comment]
@dabreegster I've rebased. I'll update the pull request description as well.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+          version: 9
+
+      - name: Install wasm-pack
+        uses: jetli/wasm-pack-action@v0.4.0
+
+      - name: Cache WASM build
+        uses: actions/cache@v3
+        with:
+          path: rust_backend/target
+          key: doesnt-matter-share-everything
```
Comment (stuartlynn) [inline .github/workflows/ci.yml:28]: Probably doesn't matter 100% but a more informative key here might be useful. I am assuming this invalidates the cache as well when rust_backend/target changes.

File: template_app/package.json [hunk 2]
```diff
@@ -6,11 +6,16 @@
     "dev": "vite dev",
     "build": "vite build",
     "preview": "vite preview",
-    "test": "npm run test:integration && npm run test:unit",
+    "test": "pnpm run test:prep && pnpm run test:all",
     "check": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json",
     "check:watch": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json --watch",
     "lint": "prettier --check . && eslint .",
```
Comment (stuartlynn) [inline template_app/package.json:12]: Perhaps worth adding a test-rust task in here to run "cargo test" in the rust lib

**Review by stuartlynn — APPROVED** [review]
Looking good. Just a few minor comments that could potentially be addressed in another PR.

Happy to merge this in when your happy with it

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  # Requires a local `pnpm` install
+  - repo: local
+    hooks:
+      - id: prettier-format
+        name: prettier-format
+        language: system
+        entry: >
+          bash -c 'pnpm run format'
+
+# sets up .pre-commit-ci.yaml to ensure pre-commit dependencies stay up to date
```
[the remark this answers is not in the document]
Comment (griff-rees) [reply .pre-commit-config.yaml]: Cool added.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (griff-rees) [reply .github/workflows/ci.yml:28]: I copied that line from ` .github/workflows/web.yml`. Happy to change to something else. Would that be applicable there as well?

File: template_app/package.json (hunk 2 above)
Comment (griff-rees) [reply template_app/package.json:12]: That looks pretty easy. Would you rather than's in `template_app/package.json` or `package.json` (or both...?).

Comment (griff-rees) [reply template_app/package.json]: Stuck it in `template_app/package.json`, so propagates to `package.json`.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -34,6 +34,11 @@ repos:
         language: system
         entry: >
           bash -c 'pnpm run format'
+      - id: cargo-format
+        name: cargor-format
```
Comment (andrewphilipsmith) [inline .pre-commit-config.yaml:38]: Is this a typo : `cargo-format` or `cargor-format`?

Comment (griff-rees) [reply .pre-commit-config.yaml]: Thanks. Resolved in 922c625ed2590bd81316123050ef8ba115427a29.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (griff-rees) [reply .github/workflows/ci.yml]: Changed to `rust_backend` in ad30590 as a demonstration @dabreegster. Totally fine to change etc.

File: .gitignore [hunk 5]
```diff
@@ -8,6 +8,9 @@ build
 !.env.example
 vite.config.js.timestamp-*
 vite.config.ts.timestamp-*
+template_app/.svelte-kit
```
Comment (dabreegster) [inline .gitignore:11]: This should be redundant with line 4, right?

**Review by dabreegster — APPROVED** [review]

File: .gitignore (hunk 5 above)
Comment (griff-rees) [reply .gitignore:11]: Ah yeah my bad

Comment (griff-rees) [reply .gitignore]: Fixed in cda69ef

**Pull request merged**